crate/operator/restore_backup.py

```diff
@@ -0,0 +1,925 @@
+        :param snapshot: The name of the snapshot to restore.
+        :param logger: the logger on which we're logging
+        """
+        try:
+            async with conn_factory() as conn:
+                async with conn.cursor() as cursor:
+                    try:
+                        await cursor.execute(
+                            "SELECT * FROM sys.snapshots WHERE repository=%s "
+                            "AND name=%s LIMIT 1;",
+                            (
+                                repository,
+                                snapshot,
+                            ),
+                        )
+                        row = await cursor.fetchone()
+                        if not row:
+                            raise kopf.PermanentError(
+                                f"Snapshot {snapshot} does not exist "
+                                f"in repository {repository}."
+                            )
+                    except ProgrammingError as e:
+                        logger.warning("Failed to get snapshots.", exc_info=e)
+        except (DatabaseError, OperationalError) as e:
+            logger.warning("DatabaseError in _ensure_snapshot_exists: %s", str(e))
+            raise kopf.TemporaryError(
+                "DatabaseError in _ensure_snapshot_exists.", delay=15
+            )
+
+    @staticmethod
+    async def _start_restore_snapshot(
+        conn_factory,
+        repository: str,
+        snapshot: str,
+        tables: List,
+        logger: logging.Logger,
+    ):
+        """
+        Run the ``RESTORE SNAPSHOT`` command to start the restore operation in the
+        target CrateDB cluster.
+
+        :param conn_factory: A function that establishes a database connection to
+            the CrateDB cluster used for SQL queries.
+        :param repository: The name of the repository.
+        :param snapshot: The name of the snapshot to restore.
+        :param tables: The list of tables that should be restored.
+        :param logger: the logger on which we're logging
+        """
+        if not tables or (len(tables) == 1 and tables[0].lower() == "all"):
+            tables_str = "all"
+        else:
+            tables_str = f'TABLE {",".join(tables)}'
+        try:
+            async with conn_factory() as conn:
+                async with conn.cursor() as cursor:
+                    repository_ident = quote_ident(repository, cursor._impl)
+                    snapshot_ident = quote_ident(snapshot, cursor._impl)
+                    try:
+                        await cursor.execute(
+                            f"RESTORE SNAPSHOT {repository_ident}.{snapshot_ident} "
+                            f"{AsIs(tables_str)} with (wait_for_completion=false)"
+                        )
+                    except DuplicateTable as e:
+                        logger.warning("Relation already exists.", exc_info=e)
+                        raise kopf.PermanentError(
```

Comment:
Does this happen if any table from the snapshot already exists? What happens to the other tables in that case?..

Reply:
yes, if at least one of the tables to restore already exists, the whole `RESTORE SNAPSHOT` operation fails.

Reply:
Cool, did not know that! We might want to explain that in the UI...